core/src/main/java/io/temporal/samples/hello/HelloSignalWithTimer.java

```diff
@@ -0,0 +1,167 @@
+  static final String WORKFLOW_ID = "HelloSignalWithTimerWorkflow";
+
+  @WorkflowInterface
+  public interface SignalWithTimerWorkflow {
+    @WorkflowMethod
+    void execute();
+
+    @SignalMethod
+    void doUpdate(String value);
+
+    @SignalMethod
+    void exit();
+  }
+
+  public static class SignalWithTimerWorkflowImpl implements SignalWithTimerWorkflow {
+
+    private Logger logger = Workflow.getLogger(SignalWithTimerWorkflowImpl.class);
+    private List<String> updates = new ArrayList<>();
+    private CancellationScope timerScope;
+    private boolean exit = false;
+    private boolean processedLast = false;
+
+    @Override
+    public void execute() {
+      // Start timer in cancellation scope so we can cancel it on exit signal received
+      timerScope =
+          Workflow.newCancellationScope(
+              () -> {
+                try {
+                  // You can add a signal handler that updates the sleep duration
+                  // As it may change via business logic over time
+                  // For sample we just hard code it to 5 seconds
+                  Workflow.newTimer(Duration.ofSeconds(5)).get();
+                } catch (CanceledFailure e) {
+                  // Exit signal is received causing cancellation of timer scope and timer
+                  // For sample we just log it, you can handle it if needed
+                  logger.info("Timer canceled via exit signal");
+                }
+              });
+      timerScope.run();
+
+      // Process last received signal and either exit or ContinueAsNew depending if we got
+      // Exit signal or not
+      // Note you would here call an activity to process last signal value received
+      // For sample we just log it in workflow rather than a dummy activity
+      try {
+        logger.info("Workflow processing last value received: " + updates.get(updates.size() - 1));
+        processedLast = true;
+      } catch (IndexOutOfBoundsException e) {
+        logger.info("No updates received, nothing to process");
+      }
+
+      if (exit) {
+        return;
+      } else {
+        SignalWithTimerWorkflow nextRun =
+            Workflow.newContinueAsNewStub(SignalWithTimerWorkflow.class);
+        nextRun.execute();
+      }
+    }
+
+    @Override
+    public void doUpdate(String value) {
+      // Note that we can receive a signal at the same time workflow is trying to complete or
+      // ContinueAsNew. This would cause workflow task failure with UnhandledCommand
+      // in order to deliver this signal to our execution.
+      // You can choose what to do in this case depending on business logic.
+      // For this sample we just ignore it, alternative could be to process it or carry it over
+      // to the continued execution if needed.
+      if (!processedLast) {
+        updates.add(value);
+      }
+    }
+
+    @Override
+    public void exit() {
+      if (timerScope != null) {
```

Comment:
I think you have an error is the `exit` signal is sent in the first WFT, you'll have to wait through the whole timer instead of cancelling because the timer won't be created yet.

Reply:
good point

Reply:
fixed